File: array/two_pointers/0121_best_time_to_sell_stock.js
/*
You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
*/

// optimize: divide into multiple non-decreasing subarray
// two-pointers
// time complexity O(n)
// space complexity O(1)
var maxProfit = function (prices) {
  let ans = 0;
  let start = 0,
    end = 1;
  while (end < prices.length) {
    if (prices[end] <= prices[start]) start = end;
    else ans = Math.max(ans, prices[end] - prices[start]);
    end++;
  }
  return ans;
};

// brute force
// time complexity O(n)
// space complexity O(1)
var maxProfit = function (prices) {
  let ans = 0;
  for (let i = 0; i < prices.length - 1; i++) {
    for (let j = i + 1; j < prices.length; j++) {
      let diff = prices[j] - prices[i];
      ans = Math.max(ans, diff);
    }
  }
  return ans;
};

File: hash_table/pairs_problem/0121_best_time_to_sell_stock.py
# You are given an array prices where prices[i] is the price of a given stock
#  on the ith day.

# You want to maximize your profit by choosing a single day to buy one stock 
# and choosing a different day in the future to sell that stock.

# Return the maximum profit you can achieve from this transaction. If you 
# cannot achieve any profit, return 0.

# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
# LC 122 https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
# LC 123 https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/
# LC 184 https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/
# LC 309 https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/
# LC 714 https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

# advanced topic: pairs problem
# LC 1014 https://leetcode.com/problems/best-sightseeing-pair/
# LC 2815 https://leetcode.com/problems/max-pair-sum-in-an-array/
# LC 2874 https://leetcode.com/problems/maximum-value-of-an-ordered-triplet-ii/
# LC 2905 https://leetcode.com/problems/find-indices-with-index-and-value-difference-ii/

# if prices[i + 1] <= price[i], never buy prices[i]
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_p, cur_min = 0, float('inf')
        for price in prices:
            cur_min = min(cur_min, price)
            max_p = max(max_p, price - cur_min)
        return max_p